packages/keymaps/src/browser/keymaps-service.ts

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (C) 2017 Ericsson and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { inject, injectable } from 'inversify';
+import { FrontendApplicationContribution, FrontendApplication } from '@theia/core/lib/browser';
+import { Disposable, DisposableCollection, CommandRegistry, KeybindingRegistry, ILogger, ResourceProvider, Resource } from '@theia/core/lib/common';
+import { RawKeybinding, KeybindingScopes } from '@theia/core/lib/common';
+import { UserStorageUri } from '@theia/userstorage/lib/browser/';
+import URI from '@theia/core/lib/common/uri';
+import * as ajv from 'ajv';
+
+export const keymapsSchema = {
+    "type": "array",
+    "properties": {
+        "keybinding": {
+            "type": "string"
+        },
+        "command": {
+            "type": "string"
+        },
+        "context": {
+            "type": "string"
+        },
+    },
+    "required": ["command", "keybinding"],
+    "optional": [
+        "context"
+    ],
+    "additionalProperties": false
+};
+
+export const keymapsUri: URI = new URI('keymaps.json').withScheme(UserStorageUri.SCHEME);
+
+@injectable()
+export class KeymapsService implements Disposable, FrontendApplicationContribution {
+
+    protected readonly toDispose = new DisposableCollection();
+    protected readonly ajv = new ajv();
+    protected ready = false;
+    protected keymapsResource: Resource;
+
+    constructor(
+        @inject(ResourceProvider) protected readonly resourceProvider: ResourceProvider,
+        @inject(CommandRegistry) protected readonly commandRegistry: CommandRegistry,
+        @inject(KeybindingRegistry) protected readonly keyBindingRegistry: KeybindingRegistry,
+        @inject(ILogger) protected readonly logger: ILogger
+    ) {
+        this.resourceProvider(keymapsUri).then(resource => {
+            if (resource) {
+                this.keymapsResource = resource;
+                this.toDispose.push(this.keymapsResource);
+                if (this.keymapsResource.onDidChangeContents) {
+                    this.keymapsResource.onDidChangeContents(() => {
+                        this.onDidChangeKeymap();
+                    });
+                }
+
+                this.keymapsResource.readContents().then(content => {
+                    this.parseKeybindings(content);
+                });
+
+            } else {
+                this.logger.error('No resource exists for: ', keymapsUri);
```

Comment:
should go to catch

Reply:
Or just remove the error now that you mentioned it above. With the user-storage-service correctly implemented (fixed in #963) it now returns something even if the resource doesn't exist.